File: composeApp/src/androidMain/kotlin/com/humblesolutions/indsphinx/viewmodel/ResidentialFormViewModel.kt
package com.humblesolutions.indsphinx.viewmodel

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope


import com.humblesolutions.indsphinx.model.ResidentialAgreement
import com.humblesolutions.indsphinx.repository.AndroidAuthRepository
import com.humblesolutions.indsphinx.repository.BackendResidentialFormRepository
import com.humblesolutions.indsphinx.repository.BackendUserProfileRepository
import com.humblesolutions.indsphinx.usecase.SubmitResidentialFormUseCase
import com.humblesolutions.indsphinx.usecase.ValidateOccupantUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

val RESPONSIBILITIES = listOf(
    "I confirm that I have received the allocated accommodation.",
    "I will maintain cleanliness and hygiene of the flat.",
    "I will not allow unauthorized visitors.",
    "I will follow company housing policies.",
    "I will report maintenance issues through the app."
)
private const val TAG = "ResidentialFormVM"

sealed class ResidentialFormUiState {
    object Loading : ResidentialFormUiState()
    data class Ready(
        val occupantName: String,
        val empId: String,
        val flatNumber: String,
        val occupantFrom: Long,
        val occupantDocId: String,
        val flatId: String,
        val commonAmenities: List<String>,
        val roomAmenities: List<String>,
        val selectedAmenities: Set<String>,
        val checkedResponsibilities: Set<Int>,
        val termsHtml: String,
        val termsAccepted: Boolean,
        val showTermsDialog: Boolean = false,
        val isSubmitting: Boolean = false
    ) : ResidentialFormUiState() {
        val allResponsibilitiesChecked get() = checkedResponsibilities.size == RESPONSIBILITIES.size
        val canSubmit get() = allResponsibilitiesChecked && termsAccepted && !isSubmitting
    }
    data class Error(val message: String) : ResidentialFormUiState()
    object Submitted : ResidentialFormUiState()
}

class ResidentialFormViewModel(application: Application) : AndroidViewModel(application) {
    private val authRepository = AndroidAuthRepository()
    private val userProfileRepository = BackendUserProfileRepository()
    private val formRepository = BackendResidentialFormRepository()
    private val validateOccupantUseCase = ValidateOccupantUseCase(userProfileRepository)
    private val submitFormUseCase = SubmitResidentialFormUseCase(formRepository)

    private val _uiState = MutableStateFlow<ResidentialFormUiState>(ResidentialFormUiState.Loading)
    val uiState: StateFlow<ResidentialFormUiState> = _uiState.asStateFlow()

    init {
        loadForm()
    }

    private fun loadForm() {
        viewModelScope.launch {
            val uid = authRepository.getCurrentUser()?.uid ?: run {
                _uiState.value = ResidentialFormUiState.Error("Session expired. Please sign in again.")
                return@launch
            }
            try {
                val profile = validateOccupantUseCase.execute(uid)
                Log.d(TAG, "loadForm: profile loaded name=${profile.name} flatNumber='${profile.flatNumber}' flatId='${profile.flatId}' hasAccepted=${profile.hasAcceptedAgreement}")
                if (profile.hasAcceptedAgreement) {
                    _uiState.value = ResidentialFormUiState.Submitted
                    return@launch
                }
                Log.d(TAG, "loadForm: fetching amenities for flatNumber='${profile.flatNumber}'")
                val (common, room) = formRepository.getFlatAmenities(profile.flatId)
                Log.d(TAG, "loadForm: amenities fetched common=$common room=$room")
                val termsHtml = formRepository.getTermsAndConditions()
                Log.d(TAG, "loadForm: termsHtml length=${termsHtml.length}")
                _uiState.value = ResidentialFormUiState.Ready(
                    occupantName = profile.name,
                    empId = profile.empId,
                    flatNumber = profile.flatNumber,
                    occupantFrom = profile.occupantFrom,
                    occupantDocId = profile.occupantDocId,
                    flatId = profile.flatId,
                    commonAmenities = common,
                    roomAmenities = room,
                    selectedAmenities = emptySet(),
                    checkedResponsibilities = emptySet(),
                    termsHtml = termsHtml,
                    termsAccepted = false
                )
            } catch (e: Exception) {
                Log.e(TAG, "loadForm: error loading form", e)
                _uiState.value = ResidentialFormUiState.Error(e.message ?: "Failed to load form.")
            }
        }
    }

    fun toggleAmenity(amenity: String) {
        val ready = _uiState.value as? ResidentialFormUiState.Ready ?: return
        val updated = if (amenity in ready.selectedAmenities) ready.selectedAmenities - amenity
                      else ready.selectedAmenities + amenity
        _uiState.update { ready.copy(selectedAmenities = updated) }
    }

    fun toggleResponsibility(index: Int) {
        val ready = _uiState.value as? ResidentialFormUiState.Ready ?: return
        val updated = if (index in ready.checkedResponsibilities) ready.checkedResponsibilities - index
                      else ready.checkedResponsibilities + index
        _uiState.update { ready.copy(checkedResponsibilities = updated) }
    }

    fun setTermsAccepted(accepted: Boolean) {
        val ready = _uiState.value as? ResidentialFormUiState.Ready ?: return
        _uiState.update { ready.copy(termsAccepted = accepted) }
    }

    fun setShowTermsDialog(show: Boolean) {
        val ready = _uiState.value as? ResidentialFormUiState.Ready ?: return
        _uiState.update { ready.copy(showTermsDialog = show) }
    }

    fun submitForm() {
        val ready = _uiState.value as? ResidentialFormUiState.Ready ?: return
        if (!ready.canSubmit) return
        _uiState.update { ready.copy(isSubmitting = true) }
        viewModelScope.launch {
            try {
                val agreement = ResidentialAgreement(
                    occupantId = ready.occupantDocId,
                    occupantName = ready.occupantName,
                    empId = ready.empId,
                    flatNumber = ready.flatNumber,
                    flatId = ready.flatId,
                    selectedAmenities = ready.selectedAmenities.toList(),
                    termsAccepted = ready.termsAccepted,
                    submittedAt = System.currentTimeMillis()
                )
                submitFormUseCase.execute(agreement)
                _uiState.value = ResidentialFormUiState.Submitted
            } catch (e: Exception) {
                _uiState.update { ready.copy(isSubmitting = false) }
                _uiState.value = ResidentialFormUiState.Error(e.message ?: "Submission failed.")
            }
        }
    }
}

File: iosApp/iosApp/ResidentialFormViewModel.swift
import Foundation
import FirebaseFirestore

let responsibilities: [String] = [
    "I confirm that I have received the allocated accommodation.",
    "I will maintain cleanliness and hygiene of the flat.",
    "I will not allow unauthorized visitors.",
    "I will follow company housing policies.",
    "I will report maintenance issues through the app."
]

@MainActor
class ResidentialFormViewModel: ObservableObject {
    private let authRepository = IOSAuthRepository()
    private let userProfileRepository = BackendUserProfileRepository()
    private let formRepository = BackendResidentialFormRepository()

    enum State {
        case loading
        case ready(
            occupantName: String,
            empId: String,
            flatNumber: String,
            occupantFrom: Date?,
            occupantDocId: String,
            flatId: String,
            commonAmenities: [String],
            roomAmenities: [String],
            selectedAmenities: Set<String>,
            checkedResponsibilities: Set<Int>,
            termsHtml: String,
            termsAccepted: Bool,
            showTermsSheet: Bool,
            isSubmitting: Bool
        )
        case error(String)
        case submitted
    }

    @Published var state: State = .loading

    init() {
        Task { await loadForm() }
    }

    private func loadForm() async {
        guard let user = authRepository.getCurrentUser() else {
            state = .error("Session expired. Please sign in again.")
            return
        }
        do {
            let profile = try await userProfileRepository.getProfile(uid: user.uid)
            guard profile.enabled else { state = .error("Your account has been disabled."); return }
            if profile.hasAcceptedAgreement { state = .submitted; return }
            let (common, room) = try await formRepository.getFlatAmenities(flatId: profile.flatId)
            let termsHtml = try await formRepository.getTermsAndConditions()
            state = .ready(
                occupantName: profile.name,
                empId: profile.empId,
                flatNumber: profile.flatNumber,
                occupantFrom: profile.occupantFrom,
                occupantDocId: profile.occupantDocId,
                flatId: profile.flatId,
                commonAmenities: common,
                roomAmenities: room,
                selectedAmenities: [],
                checkedResponsibilities: [],
                termsHtml: termsHtml,
                termsAccepted: false,
                showTermsSheet: false,
                isSubmitting: false
            )
        } catch {
            state = .error(error.localizedDescription)
        }
    }

    private func readyValues() -> (String, String, String, Date?, String, String, [String], [String], Set<String>, Set<Int>, String, Bool, Bool, Bool)? {
        guard case .ready(let n, let e, let fn, let of, let od, let fi, let ca, let ra, let sa, let cr, let th, let ta, let st, let sub) = state else { return nil }
        return (n, e, fn, of, od, fi, ca, ra, sa, cr, th, ta, st, sub)
    }

    func toggleAmenity(_ amenity: String) {
        guard let (n, e, fn, of, od, fi, ca, ra, sa, cr, th, ta, st, sub) = readyValues() else { return }
        var updated = sa
        if updated.contains(amenity) { updated.remove(amenity) } else { updated.insert(amenity) }
        state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                       commonAmenities: ca, roomAmenities: ra, selectedAmenities: updated,
                       checkedResponsibilities: cr, termsHtml: th, termsAccepted: ta, showTermsSheet: st, isSubmitting: sub)
    }

    func toggleResponsibility(_ index: Int) {
        guard let (n, e, fn, of, od, fi, ca, ra, sa, cr, th, ta, st, sub) = readyValues() else { return }
        var updated = cr
        if updated.contains(index) { updated.remove(index) } else { updated.insert(index) }
        state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                       commonAmenities: ca, roomAmenities: ra, selectedAmenities: sa,
                       checkedResponsibilities: updated, termsHtml: th, termsAccepted: ta, showTermsSheet: st, isSubmitting: sub)
    }

    func setTermsAccepted(_ accepted: Bool) {
        guard let (n, e, fn, of, od, fi, ca, ra, sa, cr, th, _, st, sub) = readyValues() else { return }
        state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                       commonAmenities: ca, roomAmenities: ra, selectedAmenities: sa,
                       checkedResponsibilities: cr, termsHtml: th, termsAccepted: accepted, showTermsSheet: st, isSubmitting: sub)
    }

    func setShowTermsSheet(_ show: Bool) {
        guard let (n, e, fn, of, od, fi, ca, ra, sa, cr, th, ta, _, sub) = readyValues() else { return }
        state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                       commonAmenities: ca, roomAmenities: ra, selectedAmenities: sa,
                       checkedResponsibilities: cr, termsHtml: th, termsAccepted: ta, showTermsSheet: show, isSubmitting: sub)
    }

    func submitForm() {
        guard let (n, e, fn, of, od, fi, ca, ra, sa, cr, th, ta, st, _) = readyValues() else { return }
        guard ta && cr.count == responsibilities.count else { return }
        state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                       commonAmenities: ca, roomAmenities: ra, selectedAmenities: sa,
                       checkedResponsibilities: cr, termsHtml: th, termsAccepted: ta, showTermsSheet: st, isSubmitting: true)
        Task {
            do {
                try await formRepository.submitAgreement(
                    occupantDocId: od, occupantName: n, empId: e, flatNumber: fn, flatId: fi,
                    selectedAmenities: Array(sa), termsAccepted: true
                )
                state = .submitted
            } catch {
                state = .ready(occupantName: n, empId: e, flatNumber: fn, occupantFrom: of, occupantDocId: od, flatId: fi,
                               commonAmenities: ca, roomAmenities: ra, selectedAmenities: sa,
                               checkedResponsibilities: cr, termsHtml: th, termsAccepted: ta, showTermsSheet: st, isSubmitting: false)
            }
        }
    }
}
